File: python/gateway/app/services/services.py

from config.config import Config
from services.character import CharacterService
from services.location import LocationService
from services.episode import EpisodeService
from model.character import CharacterClientResponse, CharacterResponse,CharacterInput,CharactersResponse
from model.episode import Episode
from model.location import Location
from typing import List,Tuple
import asyncio

def map_related_data(character, debuts, episodes, locations):
    return {
        "debut": debuts.get(character.id),
        "episodes": episodes.get(character.id),
        "origin_name": locations.get(character.origin_id) if character.origin_id is not None else None,
        "location_name": locations.get(character.location_id) if character.location_id is not None else None,
    }

class Services:
    def __init__(self, config:Config,):
        self.__character_service = CharacterService(config)
        self.__location_service = LocationService(config)
        self.__episode_service = EpisodeService(config)

    async def __fetch_related_data(self,chars: List[CharacterClientResponse]) -> Tuple[dict[int,Episode],dict[int,List[int]],dict[int,Location]]:
        ids = [x.id for x in chars]
        loc_ids = list({id for c in chars for id in (c.origin_id, c.location_id)})

        return await asyncio.gather(
            self.__episode_service.get_characters_debut_by_ids(ids),
            self.__episode_service.get_episodes_by_ids(ids),
            self.__location_service.get_locations_by_ids(loc_ids),
        )
        

    async def get_characters(self, limit: int, offset: int) -> CharactersResponse:
       

        chars, total =  await asyncio.gather(
            self.__character_service.get_characters(limit, offset),
            self.__character_service.get_total()
        )
        debuts, episodes, locations = await self.__fetch_related_data(chars)


        res_chars= []
        for c in chars:
            res_chars.append(CharacterResponse(
                id=c.id,
                name=c.name,
                status=c.status,
                species=c.species,
                type=c.type,
                gender=c.gender,
                image=c.image,
                url=c.url,
                created=c.created,
                **map_related_data(c, debuts, episodes, locations),
            ))

            print(res_chars)
        return CharactersResponse(
            total=total,
            characters= res_chars
        )
    
    
       
    

    async def get_character_by_id(self, id: int) -> CharacterResponse:
        char = await self.__character_service.get_character_by_id(id)

        debuts, episodes, locations = await self.__fetch_related_data([char])
        
        return CharacterResponse(
                id=char.id,
                name=char.name,
                status=char.status,
                species=char.species,
                type=char.type,
                gender=char.gender,
                image=char.image,
                url=char.url,
                created=char.created,
                **map_related_data(char, debuts, episodes, locations),

            )
    

    async def upsert_character(self, character: CharacterInput) -> CharacterResponse:
        return await self.__character_service.upsert_character(character)
    
    async def delete_character(self, id: int) -> dict[str,int]:
        return await self.__character_service.delete_character(id)
    

    

File: ts/gateway/src/services/services.ts
import { Config } from "../config/config.ts";
import {
  GetPageOfCharactersOptions,
  CharacterCrud,
  Character,
} from "../types/characters.ts";
import { Episode } from "../types/episode.ts";
import { Location, validateLocation } from "../types/location.ts";
import { compileCharacters } from "../utils/utils.ts";
export class Services {
  private config: Config;
  private apiUrl: string;

  constructor(config: Config) {
    this.config = config;
    this.apiUrl = config.crudServicsHost;
  }

  private async getDebutes({
    ids,
  }: {
    ids: number[];
  }): Promise<Record<number, Episode>> {
    const locationsRes = await fetch(`${this.apiUrl}/character/list/debut`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ ids }),
    });

    if (!locationsRes.ok) {
      throw new Error(`Failed to fetch episodes: ${locationsRes.statusText}`);
    }

    return await locationsRes.json();
  }
  private async getLocations({
    ids,
  }: {
    ids: number[];
  }): Promise<Record<number, Location>> {
    if (ids.length == 0) {
      return {};
    }
    const locationsRes = await fetch(`${this.apiUrl}/location/list`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ ids }),
    });

    if (!locationsRes.ok) {
      throw new Error(`Failed to fetch episodes: ${locationsRes.statusText}`);
    }

    const locs: Location[] = await locationsRes.json();
    const res: { [id: string]: Location } = {};
    locs.forEach((loc) => (res[loc.id] = loc));

    return res;
  }
  private async getEpisodes({
    ids,
  }: {
    ids: number[];
  }): Promise<Record<number, number[]>> {
    const episodesRes = await fetch(`${this.apiUrl}/character/list/episodes`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ ids }),
    });

    if (!episodesRes.ok) {
      throw new Error(`Failed to fetch episodes: ${episodesRes.statusText}`);
    }

    return await episodesRes.json();
  }
  async getCharacters(ops: GetPageOfCharactersOptions): Promise<Character[]> {
    const charsRes = await fetch(
      `${this.apiUrl}/character?limit=${ops.limit}&offset=${ops.offset}`
    );

    if (!charsRes.ok) {
      throw new Error("Failed to fetch characters");
    }

    const chars: CharacterCrud[] = await charsRes.json();

    return await compileCharacters(
      chars,
      this.getEpisodes.bind(this),
      this.getLocations.bind(this),
      this.getDebutes.bind(this)
    );
  }

  async getCharacterByID(id: number): Promise<Character | null> {
    const charRes = await fetch(`${this.apiUrl}/character/${id}`);
    if (!charRes.ok) {
      throw new Error("Failed to fetch characters");
    }

    if (charRes.body) {
      const char: CharacterCrud = await charRes.json();

      const rs = await compileCharacters(
        [char],
        this.getEpisodes.bind(this),
        this.getLocations.bind(this),
        this.getDebutes.bind(this)
      );

      if (rs.length > 0) {
        return rs[0];
      }
    }
    return null;
  }
  async upsertCharacter(char: Character): Promise<Character | null> {
    const { origin, location } = char;
    if (!validateLocation(origin)) {
      char.origin = undefined;
    }

    if (!validateLocation(location)) {
      char.location = undefined;
    }
    const upsertRes = await fetch(`${this.apiUrl}/character`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(char),
    });

    if (!upsertRes.ok) {
      throw new Error(
        `Failed to upsert character episodes: ${upsertRes.statusText}`
      );
    }

    if (upsertRes.body) {
      const char = await upsertRes.json();

      const rs = await compileCharacters(
        [char],
        this.getEpisodes.bind(this),
        this.getLocations.bind(this),
        this.getDebutes.bind(this)
      );

      if (rs.length > 0) {
        return rs[0];
      }
    }
    return null;
  }

  async deleteCharacter(id: number): Promise<{ id: number }> {
    const res = await fetch(`${this.apiUrl}/character/${id}`, {
      method: "DELETE",
      headers: {
        "Content-Type": "application/json",
      },
    });

    if (!res.ok) {
      throw new Error(`Failed to fetch episodes: ${res.statusText}`);
    }

    return { id };
  }

  async total(): Promise<number> {
    const res = await fetch(`${this.apiUrl}/total`, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
    });

    if (!res.ok) {
      throw new Error(`Failed to get total: ${res.statusText}`);
    }

    const t: { total: number } = await res.json();
    return t.total;
  }
}
